great_expectations_provider/operators/great_expectations.py

```diff
@@ -251,48 +252,97 @@ def make_connection_string(self) -> str:
-            )
-            snowflake_warehouse = (
-                self.conn.extra_dejson.get("warehouse") or self.conn.extra_dejson["extra__snowflake__warehouse"]
-            )
-            snowflake_role = self.conn.extra_dejson.get("role") or self.conn.extra_dejson["extra__snowflake__role"]
+            try:
+                from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
+                from cryptography.hazmat.backends import default_backend
+                from cryptography.hazmat.primitives import serialization
+
+                hook = SnowflakeHook(snowflake_conn_id=self.conn_id)
+
+                # Support the operator overriding the schema
+                # which is necessary for temp tables.
+                hook.schema = self.schema or hook.schema
+
+                conn = hook.get_connection(self.conn_id)
+                engine = hook.get_sqlalchemy_engine()
+
+                url = engine.url.render_as_string(hide_password=False)
+
+                private_key_file = conn.extra_dejson.get(
+                    "extra__snowflake__private_key_file"
+                ) or conn.extra_dejson.get("private_key_file")
+
+                if private_key_file:
+                    private_key_pem = Path(private_key_file).read_bytes()
+
+                    passphrase = None
+                    if conn.password:
+                        passphrase = conn.password.strip().encode()
+
+                    p_key = serialization.load_pem_private_key(
+                        private_key_pem, password=passphrase, backend=default_backend()
+                    )
+
+                    pkb = p_key.private_bytes(
+                        encoding=serialization.Encoding.DER,
+                        format=serialization.PrivateFormat.PKCS8,
+                        encryption_algorithm=serialization.NoEncryption(),
+                    )
+                    return {
+                        # Unfortunately GE uses deepcopy when instantiating the SqlAlchemyExecutionEngine
+                        # which uses pickle and SAEngine is not pickleable.
+                        # "engine": engine,
+                        "url": url,
+                        "connect_args": {
+                            "private_key": pkb,
+                        },
+                    }
 
-            uri_string = f"snowflake://{self.conn.login}:{self.conn.password}@{snowflake_account}.{snowflake_region}/{snowflake_database}/{self.schema}?warehouse={snowflake_warehouse}&role={snowflake_role}"  # noqa
+                return {"url": url}
+
+            except ImportError:
+                self.log.warning(
+                    (
+                        "Snowflake provider package not available, "
+                        "attempt to manually build connection. "
+                        "Key-based auth is not supported."
+                    )
+                )
```

Comment:
If the Snowflake Provider isn't provided, a Snowflake connection can't be made, so the conn_type should not possibly be able to be Snowflake anyway.

This import error is also misleading, because the import error could also be one of the cryptography packages, which should be specified somewhere that they're needed.

Reply:
I left out the other packages from the warning message because `cryptography` is a snowflake dependency.

It is technically possible to create a connection of any type without using its provider's package like so:
```python
from airflow.models.connection import Connection

connection = Connection(
  conn_type="snowflake",
  conn_id="foo",
  # ..etc
)
```

but, if you want to assume the provider package must be available.  I can reduce the code and remove the entire `except` block with the existing behaviour. Please let me know what you want me to do.

Reply:
Let's keep the try/except because of the example you provided, I'm convinced it's worthwhile. I just want to make sure that the block is in the right place, I still think it's incorrect to go into the try/except block first then ask whether the user is using a private key or not.

The ImportError warning message could also be a bit more clear: `attempt to manually build connection.` could be something like `attempting to build connection uri from connection components...`